### ISSUE TITLE
False-positive PossiblyNullReference

### ISSUE THREAD

**Issue by SCIF** [vimeo/psalm#5681]
https://psalm.dev/r/1436a48005

**Comment by weirdan** [comment]
That's just a syntax Psalm does not understand - asserts cannot be attached to method results. You can get your desired behaviour with a template though: https://psalm.dev/r/16bcbcf20e

**Comment by SCIF** [comment]
Hey @weirdan, thanks for your insight and workaround!

However, there are two questions still left regarding my example:
1) Why there is a test using similar syntax? https://github.com/vimeo/psalm/blob/master/tests/AssertAnnotationTest.php#L1008-L1041
2) You can see a trace saying `INFO: Trace - 25:1 - $c: A` so at least inconsistency is there

**Comment by weirdan** [comment]
> You can see a trace saying INFO: Trace - 25:1 - $c: A so at least inconsistency is there

Trace shows the type variable has *after* executing the statement the trace is attached to. So after `$c->test()` `$c` is certainly `A` - otherwise it would have crashed. To see the type *before* that statement is executed add a semicolon (empty statement):
```php
/** @psalm-trace $c */ ; // <== note the semicolon. This outputs `A|null`
/** @psalm-trace $c */   // <== this outputs `A`
$c->test();              // because it actually traces the state *after* this statement
```

**Comment by weirdan** [comment]
> Why there is a test using similar syntax?

Apologies, I was not aware that Psalm supported this syntax. The difference between that test and your code is that the class used by that test is marked *immutable*. Assertions on method returns for mutable objects are hard to reason about, as it's not clear how long they should hold, and when they need to be reset.

Converting the class to immutable makes Psalm happy: https://psalm.dev/r/7cd9725f20

Making your
